packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +540,693 @@ class ShellRoute extends RouteBase {
+  /// falling back to the identity hash code of the branch to ensure an ID is
+  /// always returned (needed for _RestorableRouteMatchList/RestorableValue).
+  String _branchLocationRestorationScopeId(StackedShellBranch branch) {
+    return branch.restorationScopeId != null
+        ? '${branch.restorationScopeId}-location'
+        : identityHashCode(branch).toString();
+  }
+
+  _RestorableRouteMatchList _branchLocation(StackedShellBranch branch) {
+    return _branchLocations.putIfAbsent(branch, () {
+      final _RestorableRouteMatchList branchLocation =
+          _RestorableRouteMatchList(_matcher);
+      registerForRestoration(
+          branchLocation, _branchLocationRestorationScopeId(branch));
+      return branchLocation;
+    });
+  }
+
+  Widget? _navigatorForBranch(StackedShellBranch branch) {
+    return _branchNavigators[branch.navigatorKey];
+  }
+
+  void _setNavigatorForBranch(StackedShellBranch branch, Widget? navigator) {
+    navigator != null
+        ? _branchNavigators[branch.navigatorKey] = navigator
+        : _branchNavigators.remove(branch.navigatorKey);
+  }
+
+  RouteMatchList? _matchListForBranch(int index) =>
+      _branchLocations[_route.branches[index]]?.value;
+
+  void _updateCurrentBranchStateFromWidget() {
+    final StackedShellBranch branch = _route.branches[_routeState.currentIndex];
+    final ShellRouteContext shellRouteContext = _routeState._shellRouteContext;
+
+    /// Create an clone of the current RouteMatchList, to prevent mutations from
+    /// affecting the copy saved as the current state for this branch.
+    final RouteMatchList currentBranchLocation =
+        shellRouteContext.routeMatchList.clone();
+
+    final _RestorableRouteMatchList branchLocation = _branchLocation(branch);
+    final RouteMatchList previousBranchLocation = branchLocation.value;
+    branchLocation.value = currentBranchLocation;
+    final bool hasExistingNavigator = _navigatorForBranch(branch) != null;
+
+    /// Only update the Navigator of the route match list has changed
+    final bool locationChanged = !RouteMatchList.matchListEquals(
+        previousBranchLocation, currentBranchLocation);
+    if (locationChanged || !hasExistingNavigator) {
+      final Widget currentNavigator = shellRouteContext.buildNavigator(
+        observers: branch.observers,
+        restorationScopeId: branch.restorationScopeId,
+      );
+      _setNavigatorForBranch(branch, currentNavigator);
+    }
+  }
+
+  void _goBranch(int index) {
+    assert(index >= 0 && index < _route.branches.length);
+    final RouteMatchList? matchlist = _matchListForBranch(index);
+    if (matchlist != null && matchlist.isNotEmpty) {
+      _routeInformationProvider.value = matchlist.toPreParsedRouteInformation();
+    } else {
+      _router.go(_routeState._effectiveInitialBranchLocation(index));
+    }
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateCurrentBranchStateFromWidget();
```

Comment:
Currently, it's needed to make sure `_updateCurrentBranchStateFromWidget` is executed once before the initial build. Were it not for the calls to registerForRestoration (by `_branchLocation(StackedShellBranch branch)`), this could be done in `initState`. And actually, I think with some modification of _RestorableRouteMatchList it could actually be moved there.